src/algorithms/reference-value.ts

```diff
@@ -0,0 +1,55 @@
+import { ReferenceResult } from '../types';
+const _ = require('lodash');
+
+export function ReferenceValue(
+  x: number,
+  formulae: Array<{
+    is_reference: boolean; // true if this is a reference value
+    method: string; // the method used to calculate the reference value
+    formula: string; // the formula used to calculate the reference value
+    min: number; // the minimum value of the reference value
+    max: number; // the maximum value of the reference value
+  }>
+): ReferenceResult {
+  // find is_reference = true array element
+  const referenceArray = formulae.filter((formula) => formula.is_reference);
+
+  // find the not reference value
+  const notReferenceArray = formulae.filter(
+    (formula) => formula.is_reference === false
+  );
+
+  let rangeValue = referenceArray.filter((r) => {
+    return r.min <= x && r.max >= x;
+  });
+
+  if (rangeValue.length === 0) {
+    rangeValue = notReferenceArray.filter((r) => {
+      return r.min <= x && r.max >= x;
+    });
+  }
+
+  // if there is no range value both reference and not reference value are in the range of x then return null
+  if (rangeValue.length === 0) {
+    return null;
+  }
+
+  const chainArray = rangeValue.map((r) => {
+    const formulaResult = eval(r.formula.replace(/X/g, x.toString()));
+    return {
+      method: r.method,
+      value: formulaResult,
+    };
+  });
+
+  const result = _.maxBy(chainArray, 'value');
+
+  if (result === undefined) {
+    return null;
+  }
+
+  return {
+    max: result.value,
+    formula: result.method,
```

Comment:
We should simply return `value` and `method`, without renaming them to max and formula.

Reply:
![Screen Shot 2022-06-21 at 13 56 52](https://user-images.githubusercontent.com/18561981/174783623-131ce935-8b93-41a5-9092-7502bbd3fc2c.png)

Reply:
That is incorrect Javascript code indeed. You should do:

```js
return {
    value: result.value,
    method: result.method
}
```

or simply `return result;`

You might want to change `ReferenceResult` type accordingly as well.